jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java

```diff
@@ -317,9 +317,9 @@ public static void annotateBytecodeInAar(
     LOG(debug, "DEBUG", "nullableReturns: " + nullableReturns);
     LOG(debug, "DEBUG", "nonnullParams: " + nonnullParams);
     // Error Prone doesn't like usages of the old Java Enumerator APIs. ZipFile does not implement
-    // Iterable, and likely
-    // never will (see  https://bugs.openjdk.java.net/browse/JDK-6581715). So this seems like the
-    // best remaining way:
+    // Iterable, and likely never will (see  https://bugs.openjdk.java.net/browse/JDK-6581715).
+    // Additionally, inputZip.stream() returns a Stream<? extends ZipEntry>, and the ::iterator
+    // method has trouble handling that. So this seems like the best remaining way:
     for (ZipEntry zipEntry : inputZip.stream().collect(ImmutableList.toImmutableList())) {
```

Comment:
If we _really_ want to avoid the copy, maybe we could handle this case with an anonymous class?  Something like:

```java
for (ZipEntry zipEntry: new Iterable<ZipEntry>() {
        Iterator<ZipEntry> iterator() {
          return inputZip.stream().iterator();
        }
     }) {
  ...
}
```
Maybe we can stick a cast in there somewhere to work around the `? extends ZipEntry` type (which I guess you can't do with a method reference)?

Reply:
I don't think that works either.

The code:

```
for (ZipEntry zipEntry: new Iterable<ZipEntry>() {
        @Override public Iterator<ZipEntry> iterator() {
          return (Iterator<ZipEntry>) inputZip.stream().iterator();
        }
      }) {
    ...
}
```

Just produces:

```
/Users/lazaro/Uber/NullAway/jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java:327: warning: [unchecked] unchecked cast
          return (Iterator<ZipEntry>) inputZip.stream().iterator();
                                                                ^
  required: Iterator<ZipEntry>
  found:    Iterator<CAP#1>
  where CAP#1 is a fresh type-variable:
    CAP#1 extends ZipEntry from capture of ? extends ZipEntry
```

And obviously it's an error without the cast too.

Reply:
Yeah, this is a tricky one :(

But if you do find a way, there's one more place where it can be applied:
https://github.com/uber/NullAway/blob/7f9aacb6b6cf54e1b6765cfe87cdba20dcb58985/jar-infer/jar-infer-lib/src/test/java/com/uber/nullaway/jarinfer/AnnotationChecker.java#L57

Reply:
Sorry, totally getting nerd sniped by this and it's not that important.  Using a good-old while loop seems to compile:

```java
    Iterator<? extends ZipEntry> iter = inputZip.stream().iterator();
    while (iter.hasNext()) {
      ZipEntry zipEntry = iter.next();
```

It's annoying to not use a for-each loop, but copying to a new list just to get around the compile error is also not great.  So maybe we should go this route?

Reply:
Sure, we can do it that way :) (No strong opinion on this myself. We could use `Enumerator` too without issue and just suppress the checker, or do the copy - since it doesn't deep copy the Zip contents, just the pointers to the entries, which in a normal AAR are like one for the AndroidManifest and one for `classes.dex` ;) )

Let me just switch to while loops on these two, though.